recirq/otoc/parallel_xeb.py

```diff
@@ -13,175 +13,219 @@
+    bits_to_probabilities,
+    angles_to_fsim,
+    pauli_error_fit,
+    generic_fsim_gate,
+    cz_to_sqrt_iswap,
+)
 
 _rot_ops = [
-    cirq.X ** 0.5, cirq.PhasedXPowGate(phase_exponent=0.25, exponent=0.5),
-    cirq.Y ** 0.5, cirq.PhasedXPowGate(phase_exponent=0.75, exponent=0.5),
-    cirq.X ** -0.5, cirq.PhasedXPowGate(phase_exponent=-0.75, exponent=0.5),
-    cirq.Y ** -0.5, cirq.PhasedXPowGate(phase_exponent=-0.25, exponent=0.5),
+    cirq.X ** 0.5,
+    cirq.PhasedXPowGate(phase_exponent=0.25, exponent=0.5),
+    cirq.Y ** 0.5,
+    cirq.PhasedXPowGate(phase_exponent=0.75, exponent=0.5),
+    cirq.X ** -0.5,
+    cirq.PhasedXPowGate(phase_exponent=-0.75, exponent=0.5),
+    cirq.Y ** -0.5,
+    cirq.PhasedXPowGate(phase_exponent=-0.25, exponent=0.5),
 ]
 _rot_mats = [cirq.unitary(r) for r in _rot_ops]
-_fsim_angle_labels = ['theta', 'delta_plus', 'delta_minus_off_diag',
-                      'delta_minus_diag', 'phi']
+_fsim_angle_labels = [
+    "theta",
+    "delta_plus",
+    "delta_minus_off_diag",
+    "delta_minus_diag",
+    "phi",
+]
 
 
-def default_interaction_sequence(
-        qubits: Sequence[cirq.GridQubit]
-) -> List[Set[Tuple[cirq.GridQubit, cirq.GridQubit]]]:
-    """Group qubits on a grid into sets of pairs.
+@dataclass
+class XEBData:
+    """Class for storing the cycle-dependent fidelities and purities of an XEB experiment."""
 
-    The qubits are restricted to be physically on a square grid with distinct
-    row and column indices (not every node of the grid needs to have a
-    qubit). They are grouped into a maximum of 4 sets, each containing pairs
-    of qubits that do not intersect with each other. The method for grouping
-    is as follows:
+    fidelity_optimized: Tuple[np.ndarray, np.ndarray]
+    fidelity_optimized_fit: Tuple[np.ndarray, np.ndarray]
+    fidelity_unoptimized: Tuple[np.ndarray, np.ndarray]
+    fidelity_unoptimized_fit: Tuple[np.ndarray, np.ndarray]
+    purity: Tuple[np.ndarray, np.ndarray]
+    purity_fit: Tuple[np.ndarray, np.ndarray]
 
-    The first set groups qubits (i, j) and (i, j + 1) where i is any integer
-    and j is an even integer. The second set groups qubits (i, j) and
-    (i + 1, j) where i is an even integer and j is any integer. The third
-    set groups qubits (i, j) and (i, j + 1) where i is any integer and j is
-    an odd integer. The last set groups qubits (i, j) and (i + 1, j) where i
-    is an odd integer and j is any integer.
 
-    After the sets are built as above, any empty set is ejected.:
+@dataclass
+class ParallelXEBResults:
+    """Class for storing results of a parallel-XEB experiment."""
 
-                 Cycle 1:                            Cycle 2:
-        q00 ── q01    q02 ── q03            q00    q01    q02    q03
-                                             |      |      |      |
-        q10 ── q11    q12 ── q13            q10    q11    q12    q13
+    fitted_gates: Dict[Tuple[Tuple[int, int], Tuple[int, int]], cirq.Circuit]
```

Comment:
Is `Tuple[Tuple[int, int], Tuple[int, int]]` a pair of grid qubits? Can we use `cirq.GridQubit` https://testing.googleblog.com/2017/11/obsessed-with-primitives.html

Reply:
Yeah that's actually what I did in the past, until we run into issues with saving them and loading them back in with pickle formats... The code won't recognize the keys when Cirq gets updated. It might be an issue with pickle and maybe some other saving formats won't make it more robust. Perhaps we can leave this for later when we have a better way to save generic data?

Reply:
Oh my goodness! This is https://github.com/quantumlib/Cirq/issues/3777 which was mania to figure out. You should have opened a bug report and saved me headache :)

Reply:
Ah, I might have mentioned this somewhere and you (or maybe someone else) told me I should not save data with pickle =) So i thought this hashing thing was intentional and didn't press it further.